File: src/main/java/net/lomeli/lomlib/util/LangUtil.java
package net.lomeli.lomlib.util;

import com.google.common.base.Strings;
import com.google.common.collect.Maps;

import java.util.Map;

import net.minecraft.util.text.TextFormatting;
import net.minecraft.util.text.translation.I18n;

public class LangUtil {
    private static String chatFormat = "/&";
    private static Map<Character, TextFormatting> formatMap;

    static {
        formatMap = Maps.newHashMap();
        for (TextFormatting formatting : TextFormatting.values())
            formatMap.put(formatting.formattingCode, formatting);
    }

    public static String translate(String unlocal, Object... args) {
        if (!Strings.isNullOrEmpty(unlocal))
            return (args != null && args.length > 0) ? String.format(translate(unlocal), args) : translate(unlocal);
        return unlocal;
    }

    private static String translate(String unlocal) {
        return I18n.translateToLocal(unlocal).replaceAll(chatFormat, "\u00a7");
    }

    private static TextFormatting getFormatFromChar(char ch) {
        TextFormatting format = formatMap.get(ch);
        if (format == null)
            format = TextFormatting.RESET;
        return format;
    }
}

File: src/main/kotlin/net/lomeli/lomlib/util/LangUtil.kt
package net.lomeli.lomlib.util

import com.google.common.base.Strings
import com.google.common.collect.Maps

import net.minecraft.util.text.TextFormatting
import net.minecraft.util.text.translation.I18n

object LangUtil {
    private val chatFormat = "/&"
    private var formatMap: MutableMap<Char, TextFormatting>? = null

    init {
        formatMap = Maps.newHashMap<Char, TextFormatting>()
        for (formatting in TextFormatting.values())
            formatMap!!.put(formatting.formattingCode, formatting)
    }

    fun translate(unlocal: String, vararg args: Any): String {
        if (!Strings.isNullOrEmpty(unlocal))
            return if (args != null && args.size > 0) String.format(translate(unlocal), *args) else translate(unlocal)
        return unlocal
    }

    private fun translate(unlocal: String): String {
        return if (I18n.canTranslate(unlocal))
            I18n.translateToLocal(unlocal).replace(chatFormat.toRegex(), "\u00a7")
        else
            I18n.translateToFallback(unlocal).replace(chatFormat.toRegex(), "\u00a7")
    }

    private fun getFormatFromChar(ch: Char): TextFormatting {
        val format = formatMap!![ch]
        return format ?: TextFormatting.RESET
    }
}
